node/src/components/contract_runtime/rewards.rs

```diff
@@ -0,0 +1,413 @@
+use std::{collections::BTreeMap, ops::Range, sync::Arc};
+
+use futures::stream::{self, StreamExt as _, TryStreamExt as _};
+use num_rational::Ratio;
+use num_traits::{CheckedAdd, CheckedMul};
+
+use crate::{
+    contract_runtime::EraValidatorsRequest,
+    effect::{
+        requests::{ContractRuntimeRequest, StorageRequest},
+        EffectBuilder,
+    },
+    types::{Block, BlockHash, Chainspec},
+};
+use casper_execution_engine::core::engine_state::{self, GetEraValidatorsError};
+use casper_hashing::Digest;
+use casper_types::{EraId, ProtocolVersion, PublicKey, U512};
+
+pub(crate) trait ReactorEventT:
+    Send + From<StorageRequest> + From<ContractRuntimeRequest>
+{
+}
+
+impl<T> ReactorEventT for T where T: Send + From<StorageRequest> + From<ContractRuntimeRequest> {}
+
+struct ErasInfo(BTreeMap<EraId, EraInfo>);
+
+/// The era information needed in the rewards computation:
+#[derive(Clone)]
+struct EraInfo {
+    weights: BTreeMap<PublicKey, U512>,
+    total_weights: U512,
+    reward_per_round: Ratio<U512>,
+}
+
+#[derive(Debug)]
+pub enum RewardsError {
+    /// There were an error while trying to get the eras information.
+    PopulateError(PopulateError),
+    /// We got a block height which is not in the era range it should be in (should not happen).
+    HeightNotInEraRange(u64),
+    /// The era is not in the range we have (should not happen).
+    EraIdNotInEraRange(EraId),
+    /// The validator public key is not in the era it should be in (should not happen).
+    ValidatorKeyNotInEra(Box<PublicKey>),
+    /// We didn't have a required switch block.
+    MissingSwitchBlock(EraId),
+    /// We got an overflow while computing something.
+    ArithmeticOverflow,
+}
+
+/// We could not fetch the asked information for the given eras.
+#[derive(Debug)]
+pub enum PopulateError {
+    FailedToFetchBlock(BlockHash),
+    FailedToFetchEra(GetEraValidatorsError),
+    /// Fetching the era validators succedeed, but no info is present (should not happen).
+    /// The `Digest` is the one that was queried.
+    NoEraReturned(Digest),
+    FailedToFetchTotalSupply(engine_state::Error),
+    FailedToFetchSeigniorageRate(engine_state::Error),
+}
+
+impl ErasInfo {
+    /// `block_hashs` is an iterator over the era ID to get the information about + the block
+    /// hash to query to have such information (which may not be from the same era).
+    pub async fn populate<REv: ReactorEventT>(
+        effect_builder: EffectBuilder<REv>,
+        protocol_version: ProtocolVersion,
+        block_hashes: impl IntoIterator<Item = (EraId, BlockHash)>,
+    ) -> Result<Self, PopulateError> {
+        let eras_info = stream::iter(block_hashes)
+            .then(|(era_id, block_hash)| async move {
+                let state_root_hash = effect_builder
```

Comment:
I think there may be a conceptual problem here that I overlooked. Since the execution of a switch block increases the token supply, we need the state root hash _prior_ to that, i.e., in practice the state root hash of the block that _precedes_ the switch block. @fizyk20, is there some trick to easily getting the prestate root hash for a block?

Reply:
This is how the data is built:

```rust
.map(|block| {
    (
        // The validator info for a switch block is the one from the next era:
        if block.is_switch() {
            block.era_id().successor()
        } else {
            block.era_id()
        },
        *block.hash(),
    )
})
```

so it should work that way. The only problem with that is genesis, but it's special cased.

Reply:
Alright, but we should add tests to this to make sure it works right